README.md
[![License: MIT](https://img.shields.io/badge/License-MIT-green.svg)](https://opensource.org/licenses/MIT)
[![Release](https://img.shields.io/github/release/emacs-grammarly/keytar-cli.svg?logo=github)](https://github.com/emacs-grammarly/keytar-cli/releases/latest)
[![npm](https://img.shields.io/npm/v/@emacs-grammarly/keytar-cli?logo=npm&color=green)](https://www.npmjs.com/package/@emacs-grammarly/keytar-cli)
[![npm-dm](https://img.shields.io/npm/dm/@emacs-grammarly/keytar-cli.svg)](https://npmcharts.com/compare/@emacs-grammarly/keytar-cli?minimal=true)

# Keytar CLI

[![CI](https://github.com/emacs-grammarly/keytar-cli/actions/workflows/test.yml/badge.svg)](https://github.com/emacs-grammarly/keytar-cli/actions/workflows/test.yml)
[![dependencies Status](https://status.david-dm.org/gh/emacs-grammarly/keytar-cli.svg)](https://david-dm.org/emacs-grammarly/keytar-cli)

keytar cli is a simple command line interface into GitHub's [node-keytar](https://github.com/atom/node-keytar) - a password/secrets management utility written for node.

## Getting Started

```
$ npx @emacs-grammarly/keytar-cli set-pass -s service1 -a testuser -p xxxxxx
Successfully set the creds for testuser@service1!

$ npx keytar-cli find-creds -s service1
Here are the creds for service1 [ { account: 'testuser', password: 'password' } ]
```
or

```
$ npm install -g @emacs-grammarly/keytar-cli
$ keytar --help
keytar is the main command, used to manage your passwords and credentials

Keytar is a native Node module to get, add, replace, and delete passwords in system's keychain.

Usage: keytar <command> [options..]

Commands:
  delete-pass [service] [account]          Delete the stored password for the `service` and `account`
  find-creds [service]                     Find all accounts and password for the `service` in the keychain
  find-pass [service]                      Find a password for the `service` in the keychain
  get-pass [service] [account]             Get the stored password for the `service` and `account`
  set-pass [service] [account] [password]  Save the `password` for the `service` and `account` to the keychain

Options:
  --version  Show version number  [boolean]
  --help     Show help            [boolean]

$ keytar set-pass -s service1 -a testuser -p xxxxxx
Successfully set the creds for testuser@service1!

$ keytar find-creds -s service1
Here are the creds for service1 [ { account: 'testuser', password: 'password' } ]
```

## License

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.


## 0.2.x (Unreleased)
> Released N/A

* Rename bin to just `keytar` (f5e2b50ef8e18ed763c2458f3ebd05ec721cc14a)
* Improve command descriptions (388da62e9fe422c3f3dee3ef258353a72d79163c)

## 0.1.4
> Released Mar 16, 2021

* First stable release.

## 0.1.x
> Released Mar 08, 2021

* Very early release.


test/commands.sh
#!/usr/bin/env bash

# Copyright (C) 2022 Jen-Chieh Shen

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3, or (at your option)
# any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with GNU Emacs; see the file COPYING.  If not, write to the
# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
# Boston, MA 02110-1301, USA.

## Commentary:
#
# Test keytar commands.
#

echo '[INFO] keytar-cli version: '
keytar --version

echo 'Set password'
keytar set-pass -s service1 -a testuser -p xxxxxx

echo 'Find credentails'
keytar find-creds -s service1


cmds/delete.js
/**
 * $File: delete.js $
 * $Date: 2021-03-08 21:52:38 $
 * $Revision: $
 * $Creator: Jen-Chieh Shen $
 * $Notice: See LICENSE.txt for modification and distribution information
 *                   Copyright © 2021 by Shen, Jen-Chieh $
 */

"use strict";

const keytar = require("keytar");

exports.command = "delete [service] [account]";
exports.desc = "Delete the stored password for the `service` and `account`.";
exports.builder = {
  service: {
    alias: "s",
    requiresArg: true
  },
  account: {
    alias: "a",
    requiresArg: true
  }
};

exports.handler = async ({ service, account }) => {
  if (!service || !account) {
    throw new Error("provide required params (service, account)");
  }
  await keytar.deletePassword(service, account);
  console.log(`Successfully delete the creds for ${account}@${service}!`);
};


cmds/get-pass.js
/**
 * $File: get-pass.js $
 * $Date: 2021-03-09 12:43:45 $
 * $Revision: $
 * $Creator: Jen-Chieh Shen $
 * $Notice: See LICENSE.txt for modification and distribution information
 *                   Copyright © 2021 by Shen, Jen-Chieh $
 */

"use strict";

const keytar = require("keytar");

exports.command = ['get-pass [service] [account]'];
exports.desc = 'get the stored password for the `service` and `account`';
exports.builder = {
  service: {
    description: 'service title',
    alias: 's',
    requiresArg: true,
    type: 'string',
  },
  account: {
    description: 'account name',
    alias: 'a',
    requiresArg: true,
    type: 'string',
  }
};

exports.handler = async ({ service, account }) => {
  if (!service || !account) {
    throw new Error("provide required params (service, account)");
  }
  const pass = await keytar.getPassword(service, account);
  console.log(pass);
};


cmds/set-pass.js
/**
 * $File: set-pass.js $
 * $Date: 2021-03-09 12:49:38 $
 * $Revision: $
 * $Creator: Jen-Chieh Shen $
 * $Notice: See LICENSE.txt for modification and distribution information
 *                   Copyright © 2021 by Shen, Jen-Chieh $
 */

"use strict";

const keytar = require("keytar");

exports.command = ['set-pass [service] [account] [password]'];
exports.desc = 'save the `password` for the `service` and `account` to the keychain';
exports.builder = {
  service: {
    description: 'service title',
    alias: 's',
    requiresArg: true,
    type: 'string',
  },
  account: {
    description: 'account name',
    alias: 'a',
    requiresArg: true,
    type: 'string',
  },
  password: {
    description: 'password or credentials',
    alias: 'p',
    requiresArg: true,
    type: 'string',
  }
};

exports.handler = async ({ service, account, password }) => {
  if (!service || !account || !password) {
    throw new Error("provide required params (service, account, password)");
  }
  await keytar.setPassword(service, account, password);
  console.log(`Successfully set the password for ${account}@${service}!`);
};


cmds/delete-pass.js
/**
 * $File: delete-pass.js $
 * $Date: 2021-03-09 12:50:16 $
 * $Revision: $
 * $Creator: Jen-Chieh Shen $
 * $Notice: See LICENSE.txt for modification and distribution information
 *                   Copyright © 2021 by Shen, Jen-Chieh $
 */

"use strict";

const keytar = require("keytar");

exports.command = ['delete-pass [service] [account]'];
exports.desc = 'delete the stored password for the `service` and `account`';
exports.builder = {
  service: {
    description: 'service title',
    alias: 's',
    requiresArg: true,
    type: 'string',
  },
  account: {
    description: 'account name',
    alias: 'a',
    requiresArg: true,
    type: 'string',
  }
};

exports.handler = async ({ service, account }) => {
  if (!service || !account) {
    throw new Error("provide required params (service, account)");
  }
  await keytar.deletePassword(service, account);
  console.log(`Successfully delete the creds for ${account}@${service}!`);
};
